### ISSUE TITLE
Lambdas in a loop (closure)

### ISSUE THREAD

**Issue by botonchou** [dennybritz/reinforcement-learning#52]
Hi, thank you. I like you codes and I'm having fun with it.

However, I noticed that in A3C sometimes some workers failed to start. I printed out the name of the worker in the [run method](https://github.com/dennybritz/reinforcement-learning/blob/master/PolicyGradient/a3c/worker.py#L95) and found out that some of the run methods were called twice, and some of them were not called at all, which led me to this [(line 125 in a3c/train.py)](https://github.com/dennybritz/reinforcement-learning/blob/master/PolicyGradient/a3c/train.py#L125)
```python
for worker in workers:
  worker_fn = lambda: worker.run(sess, coord, FLAGS.t_max)
  t = threading.Thread(target=worker_fn)
  t.start()
  worker_threads.append(t)
```
My guess is because the loop variable `worker` is not local to lambda. Please refer to [this](https://docs.python.org/3/faq/programming.html#why-do-lambdas-defined-in-a-loop-with-different-values-all-return-the-same-result) and [this](http://stackoverflow.com/questions/2295290/what-do-lambda-function-closures-capture-in-python). The reason it works for the most of the time but not always is because `t.start()` get executed quick enough before loop variable `worker` changes.

Thanks

**Comment by dennybritz** [comment]
Yeah, that's a really good catch. Thanks for reporting this. Will fix it when I have time (also feel free to submit a pull request)!

**Comment by botonchou** [comment]
Here's a simple fix:
```python
# Start worker threads
for i in range(len(workers)):
    worker_fn = lambda j=i: workers[j].run(sess, coord, FLAGS.t_max)
```
